### PR TITLE
Spatial output worker categories

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Added WorkerName category filtering to the SpatialOutput.
Use the messages from WorkerLogMessageHandler which contains the worker name to categorise worker logs.
Additionally added clearing all available log categories when running Clear Log as this gets crowded with many worker names.

Corresponding engine PR https://github.com/improbableio/UnrealEngine/pull/267

#### Release note
Feature: Added worker name categories to Spatial Output.

#### Primary reviewers
@improbable-valentyn @mattyoung-improbable

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp [hunk 1]
```diff
@@ -255,6 +264,40 @@ void SSpatialOutputLog::FormatAndPrintRawLogLine(const FString& LogLine)
...
+		WorkerNameAndType.Split(TEXT(":"), &WorkerName, &WorkerType);
+
+		// Remove the remaining '[' at the start of the WORKER_NAME.
+		WorkerName.RemoveAt(0);
+
+		// Shorten the hash at the end of the WorkerName for Unreal workers.
+		if (WorkerType.Contains("Unreal"))
+		{
+			// Keep 5 characters of the hash. e.g. UnrealWorkerF6DD366E460D1080061C2D88FFA08C1F = UnrealWorkerF6DD3.
+			WorkerName = WorkerName.Reverse();
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]:
```suggestion
			WorkerName = WorkerName.LeftChop(27);
```

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp [hunk 2]
```diff
@@ -255,6 +264,40 @@ void SSpatialOutputLog::FormatAndPrintRawLogLine(const FString& LogLine)
...
+			WorkerName.RemoveAt(0, 27); // 32 chars in hash so remove 27 from the end.
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp [hunk 3]
```diff
@@ -255,6 +264,40 @@ void SSpatialOutputLog::FormatAndPrintRawLogLine(const FString& LogLine)
...
+			WorkerName = WorkerName.Reverse();
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp [hunk 4]
```diff
@@ -255,6 +264,40 @@ void SSpatialOutputLog::FormatAndPrintRawLogLine(const FString& LogLine)
...
+		// Now remove the ':Unreal' tag
+		FString WorkerName;
+		FString WorkerType;
+		WorkerNameAndType.Split(TEXT(":"), &WorkerName, &WorkerType);
+
+		// Remove the remaining '[' at the start of the WORKER_NAME.
+		WorkerName.RemoveAt(0);
+
+		// Shorten the hash at the end of the WorkerName for Unreal workers.
+		if (WorkerType.Contains("Unreal"))
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]:
```suggestion
		if (WorkerType.Contains(TEXT("Unreal")))
```

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp [hunk 5]
```diff
@@ -125,6 +125,15 @@ void SSpatialOutputLog::OnLogDirectoryChanged(const TArray<FFileChangeData>& Fil
 	}
 }
 
+void SSpatialOutputLog::OnClearLog()
+{
+	// SOutputLog will clear the messages and the SelectedLogCategories.
+	SOutputLog::OnClearLog();
+
+	// Also clear the AvailableLogCategories as we generate many worker categories which are hard to parse.
+	Filter.AvailableLogCategories.Reset(Filter.AvailableLogCategories.GetAllocatedSize());
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]:
```suggestion
	Filter.AvailableLogCategories.Reset();
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLogMacros.cpp [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Utils/SpatialLogMacros.h"
+
+#include "Kismet/KismetSystemLibrary.h"
+
+
+DEFINE_LOG_CATEGORY(LogSpatial);
+
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialLogMacros.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp (hunk 5 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]: Based this on an optimisation in the Engine for the SelectedLogCategories.
```
void FLogFilter::ClearSelectedLogCategories()
{
	// No need to churn memory each time the selected categories are cleared
	SelectedLogCategories.Reset(SelectedLogCategories.GetAllocatedSize());
}
```

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]: That's not an optimization, it's a bug. This will increase the array's allocated buffer every time. It gets the size in bytes and tells the array to fit that many elements every time.
Here's the output I got from a simple test:
```
TArray<int> test;
test.Add(5);
for (int i = 0; i < 20; i++)
{
	UE_LOG(LogTemp, Warning, TEXT("Array allocated size: %d"), test.GetAllocatedSize());
	test.Reset(test.GetAllocatedSize());
}
```
```
[2019.10.07-20.01.26:975][477]LogTemp: Warning: Array allocated size: 16
[2019.10.07-20.01.26:975][477]LogTemp: Warning: Array allocated size: 64
[2019.10.07-20.01.26:976][477]LogTemp: Warning: Array allocated size: 256
[2019.10.07-20.01.26:976][477]LogTemp: Warning: Array allocated size: 1024
[2019.10.07-20.01.26:976][477]LogTemp: Warning: Array allocated size: 4096
[2019.10.07-20.01.26:976][477]LogTemp: Warning: Array allocated size: 16384
[2019.10.07-20.01.26:977][477]LogTemp: Warning: Array allocated size: 65536
[2019.10.07-20.01.26:977][477]LogTemp: Warning: Array allocated size: 262144
[2019.10.07-20.01.26:977][477]LogTemp: Warning: Array allocated size: 1048576
[2019.10.07-20.01.26:977][477]LogTemp: Warning: Array allocated size: 4194304
[2019.10.07-20.01.26:979][477]LogTemp: Warning: Array allocated size: 16777216
[2019.10.07-20.01.26:986][477]LogTemp: Warning: Array allocated size: 67108864
[2019.10.07-20.01.27:015][477]LogTemp: Warning: Array allocated size: 268435456
[2019.10.07-20.01.27:132][477]LogTemp: Warning: Array allocated size: 1073741824
[2019.10.07-20.01.27:610][477]LogTemp: Warning: Array allocated size: 0
[2019.10.07-20.01.27:610][477]LogTemp: Warning: Array allocated size: 0
[2019.10.07-20.01.27:610][477]LogTemp: Warning: Array allocated size: 0
[2019.10.07-20.01.27:610][477]LogTemp: Warning: Array allocated size: 0
[2019.10.07-20.01.27:611][477]LogTemp: Warning: Array allocated size: 0
[2019.10.07-20.01.27:611][477]LogTemp: Warning: Array allocated size: 0
```

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp (hunk 4 above)
Comment (yumnuska) [reply SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]: Is it intentional that this ignores the existence of AIWorker, CrashBotWorker, etc (i.e. for offloading)?

Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]: Good spot, You can get a set of current worker types from: https://github.com/spatialos/UnrealGDK/blob/358a5ef8a4b2992999eb05e2a3782dc158b425ac/SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h#L199

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp (hunk 5 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]: :o

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -37,6 +37,7 @@
 #include "Utils/OpUtils.h"
 #include "Utils/SpatialMetrics.h"
 #include "Utils/SpatialMetricsDisplay.h"
+#include "Utils/SpatialLogMacros.h"
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Alphabetize please

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp [hunk 8]
```diff
@@ -255,6 +266,38 @@ void SSpatialOutputLog::FormatAndPrintRawLogLine(const FString& LogLine)
 	// Log categories take the form "improbable.deployment.InternalGameLauncher", we filter to the last category to make it more human readable.
 	LogCategory.Split(TEXT("."), nullptr, &LogCategory, ESearchCase::IgnoreCase, ESearchDir::FromEnd);
 
+	// Remove trailing whitespace.
+	LogCategory = LogCategory.TrimEnd();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]:
```suggestion
	LogCategory.TrimEndInline();
```

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp (hunk 4 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]: So to make this correct I've had to do a bit of digging and rejigging to correctly identify an editor worker for hash shortening.

Turns out the `WorkerType` is set in the creation of the `SpatialOutputDevice`, this was hardcoded to "Unreal". In order to correctly identify editor and non-editor workers I've created a constant in `SpatialConstants` ("Unreal") for non-editor and a new constant file `SpatialGDKServicesConstants` ("UnrealEditor") for editor workers. This is a little bit cumbersome but it's because the `SpatialGDKServicesModule` where this code lives cannot have a dependency on the `SpatialGDK` as this would cause a circular dependency and break Mac and iOS builds... Additionally I'd rather have the constants than hardcoding for maintainability. I'm open to suggestions here though.

File: SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp (hunk 8 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp]: Nice!

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialStatics.h [hunk 9]
```diff
@@ -60,6 +65,18 @@ class SPATIALGDK_API USpatialStatics : public UBlueprintFunctionLibrary
 	UFUNCTION(BlueprintPure, Category = "SpatialOS|Offloading", meta = (WorldContext = "WorldContextObject"))
 	static FName GetActorGroupForClass(const UObject* WorldContextObject, const TSubclassOf<AActor> ActorClass);
 
+	/**
+	 * Functionally the same as the native Unreal PrintString but also logs to the spatial runtime.
+	 */
+	UFUNCTION(BlueprintCallable, meta = (WorldContext = "WorldContextObject", CallableWithoutWorldContext, Keywords = "log print spatial", AdvancedDisplay = "2", DevelopmentOnly), Category = "Utilities|String")
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialStatics.h]: Should we change the Category of these to "Spatial|Utilities" or something?

Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialStatics.h]: Tested this, made the discoverability worse so didn't go through with the change.
